File: use-case/enrollment/enrollment.go
package enrollmentusecase

import (
	"context"
	common "github/rakadityas/course-management-system/common"
	courseDomain "github/rakadityas/course-management-system/domain/course"
	courseEnrollmentDomain "github/rakadityas/course-management-system/domain/course-enrollment"
	studentDomain "github/rakadityas/course-management-system/domain/student"
	"strconv"
)

// EnrollmentUseCaseInterface defines the interface for the EnrollmentUseCase.
type EnrollmentUseCaseItf interface {
	CourseSignUp(ctx context.Context, req CourseSignUpRequest) (CourseSignUpResp, error)
	ListCourses(ctx context.Context, studentID int64) (ListCoursesResp, error)
	CancelCourse(ctx context.Context, studentID, courseID int64) (CancelCourseResp, error)
	ListClassmates(ctx context.Context, studentID int64) (ListClassmatesResp, error)
}

type EnrollmentUseCase struct {
	studentService          studentDomain.StudentDomainItf
	courseService           courseDomain.CourseDomainItf
	courseEnrollmentService courseEnrollmentDomain.CourseEnrollmentDomainItf
}

func NewEnrollmentUseCase(studentService studentDomain.StudentDomainItf, courseService courseDomain.CourseDomainItf, courseEnrollmentService courseEnrollmentDomain.CourseEnrollmentDomainItf) EnrollmentUseCaseItf {
	return &EnrollmentUseCase{
		studentService:          studentService,
		courseService:           courseService,
		courseEnrollmentService: courseEnrollmentService,
	}
}

// CourseSignUp handles the course sign-up process.
func (enrollmentUC *EnrollmentUseCase) CourseSignUp(ctx context.Context, req CourseSignUpRequest) (CourseSignUpResp, error) {
	// Ensure the student data exists
	studentData, err := enrollmentUC.studentService.GetStudentByID(ctx, req.StudentID)
	if err != nil {
		return CourseSignUpResp{Status: common.StatusFailure, Message: "failed to retrieve student data"}, err
	}
	if studentData == nil {
		return CourseSignUpResp{Status: common.StatusFailure, Message: "student data not found"}, nil
	}

	// Ensure the course data exists
	courseData, err := enrollmentUC.courseService.GetCourseByID(ctx, req.CourseID)
	if err != nil {
		return CourseSignUpResp{Status: common.StatusFailure, Message: "failed to retrieve course data"}, err
	}
	if courseData == nil {
		return CourseSignUpResp{Status: common.StatusFailure, Message: "course data not found"}, nil
	}

	// Ensure the student never made any enrollment at all
	courseEnrollments, err := enrollmentUC.courseEnrollmentService.GetEnrollmentByStudentIDAndCourseID(ctx, req.StudentID, req.CourseID)
	if err != nil {
		return CourseSignUpResp{Status: common.StatusFailure, Message: "failed to retrieve course data"}, err
	}
	if len(courseEnrollments) > 0 {
		return CourseSignUpResp{Status: common.StatusFailure, Message: "student has enrolled before"}, err
	}

	// Create new enrollment
	newEnrollment, err := enrollmentUC.courseEnrollmentService.CreateEnrollment(ctx, req.StudentID, req.CourseID, courseEnrollmentDomain.StatusActive)
	if err != nil {
		return CourseSignUpResp{Status: common.StatusFailure, Message: "failed to sign up course"}, err
	}

	// Return a successful response
	return CourseSignUpResp{
		Status: common.StatusSuccess,
		EnrollmentData: &CourseEnrollment{
			ID:           newEnrollment.ID,
			StudentID:    newEnrollment.StudentID,
			StudentEmail: studentData.Email,
			CourseID:     newEnrollment.CourseID,
			CourseName:   courseData.Name,
			Status:       courseEnrollmentDomain.StatusActive,
			CreateTime:   newEnrollment.CreateTime,
			UpdateTime:   newEnrollment.UpdateTime,
		},
	}, nil
}

// ListCourses retrieves the list of courses a student is enrolled in.
func (enrollmentUC *EnrollmentUseCase) ListCourses(ctx context.Context, studentID int64) (ListCoursesResp, error) {
	// Ensure the student data exists
	studentData, err := enrollmentUC.studentService.GetStudentByID(ctx, studentID)
	if err != nil {
		return ListCoursesResp{Status: common.StatusFailure, Message: "failed to retrieve student data"}, err
	}
	if studentData == nil {
		return ListCoursesResp{Status: common.StatusFailure, Message: "student data not found"}, nil
	}

	// Get course enrollments for the student
	enrollments, err := enrollmentUC.courseEnrollmentService.GetEnrollmentByStudentID(ctx, studentID)
	if err != nil {
		return ListCoursesResp{Status: common.StatusFailure, Message: "failed to retrieve enrollments"}, err
	}

	// Prepare the response
	var courses []CourseDetail
	for _, enrollment := range enrollments {
		course, err := enrollmentUC.courseService.GetCourseByID(ctx, enrollment.CourseID)
		if err != nil {
			return ListCoursesResp{Status: common.StatusFailure, Message: "failed to retrieve course data"}, err
		}
		if course == nil {
			return ListCoursesResp{Status: common.StatusFailure, Message: "course data is not found for courseID: " + strconv.FormatInt(enrollment.CourseID, 10)}, nil
		}

		courses = append(courses, CourseDetail{
			CourseID:   course.ID,
			CourseName: course.Name,
			Status:     enrollment.Status,
			CreateTime: enrollment.CreateTime,
			UpdateTime: enrollment.UpdateTime,
		})
	}

	return ListCoursesResp{
		Status:  common.StatusSuccess,
		Courses: courses,
	}, nil
}

// CancelCourse cancel registered courses on the course enrollment table
func (enrollmentUC *EnrollmentUseCase) CancelCourse(ctx context.Context, studentID, courseID int64) (CancelCourseResp, error) {
	err := enrollmentUC.courseEnrollmentService.UpdateCourseEnrollmentStatus(ctx, studentID, courseID, courseEnrollmentDomain.StatusCancelled)
	if err != nil {
		return CancelCourseResp{Status: common.StatusFailure, Message: "failed to cancel course enrollment"}, err
	}

	return CancelCourseResp{
		Status: common.StatusSuccess,
	}, nil
}

// ListClassmates retrieves the list of classmates for the given student.
func (enrollmentUC *EnrollmentUseCase) ListClassmates(ctx context.Context, studentID int64) (ListClassmatesResp, error) {
	// Ensure the student data exists
	studentData, err := enrollmentUC.studentService.GetStudentByID(ctx, studentID)
	if err != nil {
		return ListClassmatesResp{Status: common.StatusFailure, Message: "failed to retrieve student data"}, err
	}
	if studentData == nil {
		return ListClassmatesResp{Status: common.StatusFailure, Message: "student data not found"}, nil
	}

	// Get course enrollments for the student
	enrollments, err := enrollmentUC.courseEnrollmentService.GetListClassmates(ctx, studentID)
	if err != nil {
		return ListClassmatesResp{Status: common.StatusFailure, Message: "failed to get list of classmates"}, err
	}

	// Create a map to group students by course ID
	mapCourseGroup := make(map[int64][]int64)
	for _, enrollment := range enrollments {
		mapCourseGroup[enrollment.CourseID] = append(mapCourseGroup[enrollment.CourseID], enrollment.StudentID)
	}

	// Prepare the response
	var response ListClassmatesResp
	for courseID, studentIDs := range mapCourseGroup {
		course, err := enrollmentUC.courseService.GetCourseByID(ctx, courseID) // todo: improve this with get bulk
		if err != nil {
			return ListClassmatesResp{Status: common.StatusFailure, Message: "failed to retrieve course data"}, err
		}
		if course == nil {
			return ListClassmatesResp{Status: common.StatusFailure, Message: "course data is not found for courseID: " + strconv.FormatInt(courseID, 10)}, nil
		}

		var classmates []ListClassmatesStudentsResp
		for _, id := range studentIDs {
			if id == studentID {
				continue // Skip the current student
			}

			student, err := enrollmentUC.studentService.GetStudentByID(ctx, id) // todo: improve this with get bulk
			if err != nil {
				return ListClassmatesResp{Status: common.StatusFailure, Message: "failed to retrieve student data: " + strconv.FormatInt(id, 10)}, err
			}
			if student == nil {
				return ListClassmatesResp{Status: common.StatusFailure, Message: "student data is not found for studentID: " + strconv.FormatInt(id, 10)}, nil
			}

			classmates = append(classmates, ListClassmatesStudentsResp{
				StudentID:    strconv.FormatInt(student.ID, 10),
				StudentEmail: student.Email,
			})
		}

		response.Courses = append(response.Courses, ListClassmatesCourseResp{
			CourseID:   course.ID,
			CourseName: course.Name,
			ClassMates: classmates,
		})
	}

	return ListClassmatesResp{
		Status:  common.StatusSuccess,
		Courses: response.Courses,
	}, nil
}

File: python/use_case/enrollment/enrollment.py
from typing import List
from .types import (
    CourseSignUpRequest,
    CourseSignUpResp,
    CourseEnrollmentResp,
    ListCoursesResp,
    CourseDetail,
    CancelCourseResp,
    ListClassmatesResp,
    ListClassmatesCourseResp,
    ListClassmatesStudentsResp,
)
from python.common.const import StatusFailure, StatusSuccess
from python.domain.student.service import StudentService
from python.domain.course.service import CourseService
from python.domain.course_enrollment.service import CourseEnrollmentService
from python.domain.course_enrollment.models import StatusActive, StatusCancelled

class EnrollmentUseCase:
    def __init__(self, student_service: StudentService, course_service: CourseService, course_enrollment_service: CourseEnrollmentService):
        self.student_service = student_service
        self.course_service = course_service
        self.course_enrollment_service = course_enrollment_service

    def course_sign_up(self, req: CourseSignUpRequest) -> CourseSignUpResp:
        student = self.student_service.get_student_by_id(req.student_id)
        if student is None:
            return CourseSignUpResp(status=StatusFailure, message="student data not found")

        course = self.course_service.get_course_by_id(req.course_id)
        if course is None:
            return CourseSignUpResp(status=StatusFailure, message="course data not found")

        existing = self.course_enrollment_service.get_enrollment_by_student_and_course(req.student_id, req.course_id)
        if len(existing) > 0:
            return CourseSignUpResp(status=StatusFailure, message="student has enrolled before")

        new_enrollment = self.course_enrollment_service.create_enrollment(req.student_id, req.course_id, StatusActive)
        return CourseSignUpResp(
            status=StatusSuccess,
            enrollment_data=CourseEnrollmentResp(
                id=new_enrollment.id,
                student_id=new_enrollment.student_id,
                student_email=student.email,
                course_id=new_enrollment.course_id,
                course_name=course.name,
                status=StatusActive,
                create_time=new_enrollment.create_time,
                update_time=new_enrollment.update_time,
            ),
        )

    def list_courses(self, student_id: int) -> ListCoursesResp:
        student = self.student_service.get_student_by_id(student_id)
        if student is None:
            return ListCoursesResp(status=StatusFailure, message="student data not found")

        enrollments = self.course_enrollment_service.get_enrollment_by_student_id(student_id)
        courses: List[CourseDetail] = []
        for e in enrollments:
            course = self.course_service.get_course_by_id(e.course_id)
            if course is None:
                return ListCoursesResp(status=StatusFailure, message=f"course data is not found for courseID: {e.course_id}")
            courses.append(
                CourseDetail(
                    course_id=course.id,
                    course_name=course.name,
                    status=e.status,
                    create_time=e.create_time,
                    update_time=e.update_time,
                )
            )
        return ListCoursesResp(status=StatusSuccess, courses=courses)

    def cancel_course(self, student_id: int, course_id: int) -> CancelCourseResp:
        updated = self.course_enrollment_service.update_course_enrollment_status(student_id, course_id, StatusCancelled)
        if not updated:
            return CancelCourseResp(status=StatusFailure, message="failed to cancel course enrollment")
        return CancelCourseResp(status=StatusSuccess)

    def list_classmates(self, student_id: int) -> ListClassmatesResp:
        student = self.student_service.get_student_by_id(student_id)
        if student is None:
            return ListClassmatesResp(status=StatusFailure, message="student data not found")
        enrollments = self.course_enrollment_service.get_list_classmates(student_id)
        grouped = {}
        for e in enrollments:
            grouped.setdefault(e.course_id, []).append(e.student_id)
        courses_resp = []
        for course_id, student_ids in grouped.items():
            course = self.course_service.get_course_by_id(course_id)
            if course is None:
                return ListClassmatesResp(status=StatusFailure, message=f"course data is not found for courseID: {course_id}")
            classmates = []
            for sid in student_ids:
                if sid == student_id:
                    continue
                s = self.student_service.get_student_by_id(sid)
                if s is None:
                    return ListClassmatesResp(status=StatusFailure, message=f"student data is not found for studentID: {sid}")
                classmates.append(ListClassmatesStudentsResp(student_id=str(s.id), student_email=s.email))
            courses_resp.append(ListClassmatesCourseResp(course_id=course.id, course_name=course.name, class_mates=classmates))
        return ListClassmatesResp(status=StatusSuccess, courses=courses_resp)
